substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+			}
+
+			Ok(Pays::No.into())
+		}
+
+		/// Plan an epoch config change.
+		///
+		/// The epoch config change is recorded and will be announced at the begin of the
+		/// next epoch together with next epoch authorities information.
+		/// In other words the configuration will be activated one epoch after.
+		/// Multiple calls to this method will replace any existing planned config change that had
+		/// not been enacted yet.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::plan_config_change())]
+		pub fn plan_config_change(
+			origin: OriginFor<T>,
+			config: EpochConfiguration,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			ensure!(
+				config.redundancy_factor != 0 && config.attempts_number != 0,
+				Error::<T>::InvalidConfiguration
+			);
+			PendingEpochConfigChange::<T>::put(config);
+			Ok(())
+		}
+	}
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
+		type Call = Call<T>;
+
+		fn validate_unsigned(source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			let Call::submit_tickets { tickets } = call else {
+				return InvalidTransaction::Call.into()
+			};
+
+			// Discard tickets not coming from the local node or that are not
+			// yet included in a block
+			debug!(
+				target: LOG_TARGET,
+				"Validating unsigned from {} source",
+				match source {
+					TransactionSource::Local => "local",
+					TransactionSource::InBlock => "in-block",
+					TransactionSource::External => "external",
+				}
+			);
+
+			if source == TransactionSource::External {
+				// TODO @davxy: BRAINSTORM this `Local` requirement...
+				// If we only allow these txs on block production, then there is less chance to
+				// submit our tickets if we don't have enough authoring slots.
+				// If we have 0 slots => we have zero chances.
+				// Maybe this is one valid reason to introduce proxies.
+				// In short the question is >>> WHO HAS THE RIGHT TO SUBMIT A TICKET? <<<
+				//  A) The current epoch validators
+				//  B) Doesn't matter as far as the tickets are good (i.e. RVRF verify is ok)
+				// Maybe we also provide a signed extrinsic to submit tickets
+				// where the submitter doesn't pay if the tickets are good?
+				warn!(
+					target: LOG_TARGET,
+					"Rejecting unsigned `submit_tickets` transaction from an external source",
+				);
+				return InvalidTransaction::BadSigner.into()
+			}
+
+			// Current slot should be less than half of epoch length.
+			let epoch_length = T::EpochLength::get();
+
+			let current_slot_idx = Self::current_slot_index();
+			if current_slot_idx > epoch_length / 2 {
+				warn!(target: LOG_TARGET, "Timeout to propose tickets, bailing out.",);
+				return InvalidTransaction::Stale.into()
+			}
+
+			// This should be set such that it is discarded after the first epoch half
+			let tickets_longevity = epoch_length / 2 - current_slot_idx;
+			let tickets_tag = tickets.using_encoded(|bytes| hashing::blake2_256(bytes));
```

Comment:
If the tickets are submitted twice, with different order then provides tag will be different. Don't know if this is problem, we could sort before hashing.

Reply:
mmm... well I thing that if the tickets have a different order this would be a different call I suppose.

Probably later we may think about adding some mechanism that keeps track of the identity of the submitter together with the body of the ticket and eventually slash if he submits duplicates.

We will also need to add a mechanism to slash submitters of invalid tickets (e.g. slash proportional to the number of invalid / duplicated tickets).

Reply:
Different call, yes, but the state change this call brings is the same. And that is why provides tag could be the same. But it is not crucial.

I have another random thought: could we assume that tickets in `submit_tickets` must be sorted and make it hard requirement?

Reply:
I drop here an additional note that may be useful in the future when considering possibilities for slashing (https://github.com/paritytech/polkadot-sdk/pull/1577#issuecomment-1812707909).

Scenario
1. Alice submits [t1, t2]
2. Bob submits [t1, t2]       <- doesn't enter the pool as `tickets_tag` is equal to the Alice one from 1
3. Bob submits [t2]             <- refused in the `submit_tickets` call but no slashing happens as this is like the first time we see Bob

But maybe instead Bob should be slashed because he submitted t2 twice.